r/src/array.cpp

```diff
@@ -0,0 +1,209 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow_types.h"
+
+using namespace Rcpp;
+using namespace arrow;
+
+namespace arrow{
+namespace r{
+
+template <int RTYPE, typename Vec = Rcpp::Vector<RTYPE> >
+class SimpleRBuffer : public arrow::Buffer {
+public:
+
+  SimpleRBuffer(Vec vec) :
+    Buffer(reinterpret_cast<const uint8_t*>(vec.begin()), vec.size() * sizeof(typename Vec::stored_type)),
+    vec_(vec)
+  {}
+
+private:
+  // vec_ holds the memory
+  Vec vec_;
+};
+
+template <int RTYPE, typename Type>
+std::shared_ptr<arrow::Array> SimpleArray(SEXP x){
+
+  Rcpp::Vector<RTYPE> vec(x);
+  std::vector<std::shared_ptr<arrow::Buffer>> buffers {
+    nullptr,
+    std::make_shared<SimpleRBuffer<RTYPE>>(vec)
+  };
+
+  int null_count = 0;
+  if (RTYPE != RAWSXP) {
+    // TODO: maybe first count NA in a first pass so that we
+    //       can allocate only if needed
+    std::shared_ptr<arrow::Buffer> null_bitmap;
+    R_ERROR_NOT_OK(arrow::AllocateBuffer(vec.size(), &null_bitmap));
+
+    auto null_bitmap_data = null_bitmap->mutable_data();
+    for (int i=0; i < vec.size(); i++) {
+      if (Rcpp::Vector<RTYPE>::is_na(vec[i]) ) {
+        BitUtil::SetBit(null_bitmap_data, i);
```

Comment:
Use FirstTimeBitmapWriter here for better performance https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit-util.h#L425